spec/lib/pedicel/base_spec.rb

```diff
@@ -0,0 +1,425 @@
+        permutated_params = permutated_keys.zip(params.values).to_h
+
+        expect{Pedicel::Base.verify_x509_chain(permutated_params)}.to raise_error(Pedicel::SignatureError, /invalid chain/)
+      end
+    end
+
+    it 'errs if intermediate equals leaf (because root did not sign leaf)' do
+      params[:intermediate] = params[:leaf]
+      expect{Pedicel::Base.verify_x509_chain(params)}.to raise_error(Pedicel::SignatureError, /invalid chain due to intermediate/)
+    end
+
+    it 'errs if leaf equals intermediate (because the intermediate must sign leaf)' do
+      params[:leaf] = params[:intermediate]
+      expect{Pedicel::Base.verify_x509_chain(params)}.to raise_error(Pedicel::SignatureError, /invalid chain/)
+    end
+
+    it 'errs if intermediate equals root (because root did not sign leaf)' do
+      params[:intermediate] = params[:root]
+      expect{Pedicel::Base.verify_x509_chain(params)}.to raise_error(Pedicel::SignatureError, /invalid chain/)
+    end
+
+    it 'errs if root equals intermediate (because intermediate is not self-signed)' do
+      params[:root] = params[:intermediate]
+      expect{Pedicel::Base.verify_x509_chain(params)}.to raise_error(Pedicel::SignatureError, /invalid chain/)
+    end
+
+    it 'errs if leaf equals root (because intermediate must sign leaf)' do
+      params[:leaf] = params[:root]
+      expect{Pedicel::Base.verify_x509_chain(params)}.to raise_error(Pedicel::SignatureError, /invalid chain/)
+    end
+
+    it 'errs if root equals leaf (becuase leaf is not self-signed)' do
+      params[:root] = params[:leaf]
+      expect{Pedicel::Base.verify_x509_chain(params)}.to raise_error(Pedicel::SignatureError, /invalid chain/)
+    end
+
+    it 'errs if leaf is used for all 3 certificates (because of multiple reasons)' do
+      params[:root]         = params[:leaf]
+      params[:intermediate] = params[:leaf]
+      expect{Pedicel::Base.verify_x509_chain(params)}.to raise_error(Pedicel::SignatureError, /invalid chain/)
+    end
+
+    it 'errs if intermediate is used for all 3 certificates' do
+      params[:root] = params[:intermediate]
+      params[:leaf] = params[:intermediate]
+      expect{Pedicel::Base.verify_x509_chain(params)}.to raise_error(Pedicel::SignatureError, /invalid chain/)
+    end
+
+    #it 'does not err when root is used for all 3 certificates' do
+    #  params[:intermediate] = params[:root]
+    #  params[:leaf]         = params[:root]
+    #  expect{Pedicel::Base.verify_x509_chain(params)}.to_not raise_error
+    #end
+    #
+    # Intentionally removed. It would fail with the current implementation.
+    #
+    # The thought behind the test is that if root was used in all three places,
+    # then all conditions are fulfilled, but it would also be quite strange to
+    # see Apple's Root CA certificate sign a payment token. Thus, we accept that
+    # this theoretically acceptable chain is not accepted (because it will never
+    # happen).
```

Comment:
Perhaps we should have the opposite test? That using root for all three certificates is explicitly disallowed. Maybe it is just a waste of time. Someone trying to trick us would still need to have Apple's Root CA sign data, i.e. it will not happen.

Reply:
Having the current comment around shows that we (you) have had a thought, which is fine :+1: